### ISSUE TITLE
Remove current Nuspell+ICU4C implementation and integrate spell check into new suggestion/autocorrect system

### ISSUE THREAD

**Issue by patrickgold** [florisboard/florisboard#1921]
This issue is primarily to keep track of the spell check re-implementation (full integration into the new 0.4 suggestion system), which is driven by 3 main factors:

1. **License incompatibility between FlorisBoard and Nuspell/Hunspell**  
   FlorisBoard's license (Apache 2.0) and Nuspell/Hunspell's license (LGPL-3.0) are somewhat incompatible with each other, which causes a license conflict. This has been found due to discussions around a potential inclusion of FlorisBoard into GrapheneOS. The only way to resolve this is to fully step away from using Hunspell altogether, as all libraries working with these dictionary formats are LGPL license variants.
2. **Quality and reliability of Hunspell dictionaries and the dictionary UI**  
   There have been numerous bugs regarding encoding issues with some Hunspell dictionaries, which were difficult to impossible to fix. These encoding issues ranged from strange corrections to no corrections being generated at all, or all words being marked as incorrect. Additionally the dictionary logic was in dire need of a rewrite anyways, as it was ported over from pre 0.3.14.
3. **Duplicate dictionary systems for next-word suggestions and spell checking**  
   When suggestions are implemented, it means that there is a new dictionary loaded for each language. So for a user to have spell check and suggestions for a particular language, at least two different dictionaries for the same language must be installed separately. In order to simplify and unify this process, Nuspell is being removed and spell check is integrated into the new NLP logic, fully covering spell check and next-word suggestions/auto-correct.

Additionally, maintainability should be easier with a unified spell check/suggestion system. Also the next/current-word suggestions have to perform autocorrect anyways, which is a form of spell checking, so a unification of these systems is the best way to go.

With the removal of Nuspell ICU4C will probably also not be needed anymore, however this is not decided yet. If it is deemed to be unnecessary, the removal of ICU4C means lifting the ABI restrictions of the APK, so stable build APKs should be automatically deployable on x86/x86_64 again, without building from the source itself (not 0.4.0-alpha1 though).

### Plan for spell check:
- **0.4.0-alpha1**  
   Fully remove spell check backend, integrate existing spell check service into the new placeholder NLP provider interface. This means that this alpha version will not have a working spell checker.
- **0.4.0-alpha2+**  
   As the suggestion implementation then gets experimented and tested out, this will also directly affect the spell check and the service will start working again depending on the state of the suggestion implementation.
- **0.4 release candidates/stable**
   Together with the suggestion implementation, spell check should be up and running flawlessly again.

**Comment by Glitchy-Tozier** [comment]
Do you know why they follow a strict no-(L)GPL policy?

**Comment by patrickgold** [comment]
See https://github.com/GrapheneOS/os-issue-tracker/issues/579#issuecomment-1156631530

**Comment by Glitchy-Tozier** [comment]
In case you're interested in my take:
As a user, I don't care that much about inclusion of FlorisBoard in GrapheneOS. I'd like for spell-checking to run as smoothly as possible, suggesting the correct words as often as possible.

If that requires sticking with Nuspell, I'd prefer it over using a worse library just for appeasing GrapheneOS.

On the other hand, if Nuspell was troublesome to work with to begin with and other libraries' **functionality** is better anyways, going the other route seems better.

**Comment by patrickgold** [comment]
Well it is both to be honest.

Nuspell itself is not that troublesome to work with, but its dependency, ICU4C. It took me a long time to even get somehow compiled for Android devices, and even now it still causes encoding issues, which result in issues such as:
- All words being marked incorrect
- Suggestions incorrect or outright blank
- Suggestions having gibberish chars
- Issues with non-Latin languages

The problem with ICU4C is, that I have to severely strip it down so my APK size does not explode, and I probably exclude something which is needed for some devices. But as my knowledge of ICU is limited, I don't really know where to start debugging and fixing these encoding issues while not increasing the APK size severely. Using the system ICU4C would solve a lot of these issues, however this is only available on Android 12+, so tbh never for this project, considering how many devices are stuck with older Android versions.

Then on the GrapheneOS inclusion itself: The reason why I am generally open for inclusion is because it would bring more attention to this project, and if this keyboard is the default for an OS also more incentives for experienced users to help contribute fixes for bugs and implementations for features.

Maybe I can make Nuspell a separate module and toggle-able during build time in the build.gradle.kts, so GrapheneOS can exclude Nuspell while this project's users can use it for the spell check service. But this still requires that I fix the encoding issues and that the module idea does not require too much extra effort to maintain.

> I'd like for spell-checking to run as smoothly as possible, suggesting the correct words as often as possible.

Yes, I agree this is a high priority that this works well. But Nuspell is single-word spell-checking only, thus I have to implement something else for auto-correction/suggestions anyways. And auto-correct and spell check goes somewhat in the same direction anyways, so I thought I combine those into one in 0.4.

**Comment by Glitchy-Tozier** [comment]
> The reason why I am generally open for inclusion is because it would bring more attention to this project, and if this keyboard is the default for an OS also more incentives for experienced users to help contribute fixes for bugs and implementations for features.

I understand that the inclusion is tempting, I'm just saying it should be thought through and not take pecedence over making the keyboard as great as possible.

> Maybe I can make Nuspell a separate module and toggle-able during build time in the build.gradle.kts, so GrapheneOS can exclude Nuspell while this project's users can use it for the spell check service. But this still requires that I fix the encoding issues and that the module idea does not require too much extra effort to maintain.
>
>>I'd like for spell-checking to run as smoothly as possible, suggesting the correct words as often as possible.
>
>Yes, I agree this is a high priority that this works well. But Nuspell is single-word spell-checking only, thus I have to implement something else for auto-correction/suggestions anyways. And auto-correct and spell check goes somewhat in the same direction anyways, so I thought I combine those into one in 0.4.

I trust that you'll make a good call!  :)
I just wanted to mention my thoughts on functionality > inclusion.

**Comment by Kreuger** [comment]
Some great discussion here guys. Just want to say that I appreciate the transparency. I fully understand both sides of this argument. And theyre both valid.

**Comment by rancidfrog** [comment]
@patrickgold 
https://github.com/bromite/bromite/issues/2141

In regards to grapheneOS,
It seems they create toxic atmosphere.

You should only collaborate with devs that act professionally.

**Comment by patrickgold** [comment]
@rancidfrog Oh good god what a disastrous drama, thanks for bringing this to my attention.

Tbh I don't really understand what the conflict is even about and I don't want to waste my time doing extensive research what happened between those two communities nor who began it all along. Like shouldn't we as the open-source community be united, so we have a chance against the big billion-dollar fishes? Instead some open-source communities fight each other for what?

I just want to develop open-source software in **peace**, I do not care **at all** for these completely unnecessary wars between communities and certainly do not want to be dragged into those. As such I also do not plan to research further into this conflict and will probably withdraw my direct support for an inclusion into GrapheneOS. This is very sad as from a technical standpoint GrapheneOS is a good project and I liked it, but it seems to be what it is. Also I want my project's climate to be friendly and welcoming, not toxic and always at war with someone just because I accepted a PR from a developer a certain community dislikes.

Nonetheless this will not affect my direct plans with spell checking above, as the current implementation has its technical flaws and I need to re-check the license compatibility between FlorisBoard and Nuspell just to make sure I do not violate the LGPL-3.0 license.

**Comment by Kreuger** [comment]
What in the ever loving hell is going on over there? I cant make heads or tails of it. Seems like a bunch of he-said/she-said.

**Comment by alexmo1997** [comment]
> Yeah I also don't know what this is exactly about but like I said I won't think too much about it and try to avoid the drama (and both parties as a result) as best as possible.

I think this doesn't really affect Florisboard, and while I don't agree with the means, there might be good reasons GrapheneOS doesn't want to continue working with the Bromite team. Thus I think this shouldn't affect the collaboration between GOS and Florisboard when it comes to inclusion of the latter as the default GOS keyboard. Both projects are currently, in my humble opinion, the best projects there are in their area so the collaboration could really help both projects.

**Comment by patrickgold** [comment]
Yeah but how can I ensure I don't get into a drama at some point? Like I already menrioned on a technical level I would have really liked the collab and it would have been beneficial to both me and GrapheneOS, but the drama is a really really big deterrent in this topic now...

**Comment by alexiaoff** [comment]
@patrickgold 
Hello, I have created my github account only you ask by message to please associate you with the GrapheneOS project. I am a user of this OS for 2 years now and since can on their matrix channel and I ask you to believe me when I tell you that the GrapheneOS team is extremely professional and attentive with its users. You have no reason to worry about your project if you work together since GrapheneOS already works with other developers and have no worries. So I ask you as a user to please work with GrapheneOS. Your two projects are exceptional.
Thank you

**Comment by unilock** [comment]
In my opinion, this drama is neither here nor there.
As an end user, I believe the project's focus should be on improving Florisboard, not complying with the arbitrary standards of third parties - much less worrying about their external affairs.

Although I do think modularizing Florisboard in general is a good idea - not just for the spell checking implementation, but for other third-party libraries as well.
(note that I have no experience with dependency management, so I may be underestimating a herculean task)

In any case - this is getting pretty off-topic, isn't it? :)

(No affiliation with GrapheneOS nor Bromite.)

**Comment by Dyras** [comment]
@alexiaoff My 2 cents in this discussion is that if anything, Patrick Gold and Carl Sagan are the professionals here. I see no reason for why the lead developer of GrapheneOS, Daniel Micay, should go out of his way to attack someone over a PR, simply because that person is (apparently?) affiliated with CalyxOS.

I've followed the GrapheneOS and CalyxOS drama for years now, and I still don't know what exactly sparked all this fire. I've seen zero screenshots of what this so called misinformation is. On the contrary, it seems Micay is searching for reasons to fight with people. This is hardly the first time he's gone on rants for little to no reason with someone online.

Which is a pity, really. As Patrick Gold mentions, I feel that open source communities should stick together in the great battle against big tech instead of fighting over pull requests.

But yeah, this is off-topic etc. And Gold has my full support if he doesn't want drama in his life. I wouldn't either in his shoes. Discussions about licensing is one thing, but straight out threats like those done against Carl Sagan... that's just taking it too far.

**Comment by alexiaoff** [comment]
@unilock 
Arbitrary standards? There is talk of a license that would allow manufacturers to use florisboard, this is not off topic. As a user I would be happy if the OS I use used florisboard as default keyboard.

**Comment by alexiaoff** [comment]
@Dyras 
It was you and a few others who came to bring your drama as you say. At the beginning it was an inclusion of florisboard in graphemes until YOUR drama came to pollute

**Comment by alexiaoff** [comment]
It is you and only you who come here to talk about bromite and calyxos. 
I as a user florisboard be integrated with grapheneos.

**Comment by Dyras** [comment]
@alexiaoff Me? What have I done? All I'm saying is that this all happened because Daniel Micay jumped Carl Sagan over a pull request. Why would Patrick Gold want to be associated with a project that fights people over pull requests? Why would he want to risk that drama?

Literally all Micay had to do was stay quiet and not stick his nose in other people's projects and we would see Florisboard in GrapheneOS by the end of the year.

**Comment by alexiaoff** [comment]
@Dyras 
And if you want to support florisboard why not encourage it to become the default keyboard in OS?

**Comment by alexiaoff** [comment]
@Dyras 
But what do we care about bromite, here we discuss florisboard! Stop with your misinformation

**Comment by alexiaoff** [comment]
You do not support any project by preventing florisboard from being integrated into an OS.

**Comment by Dyras** [comment]
@alexiaoff 
What I want is for the developer of this great app to not be the target of drama in case he considers a Pull Request from a person associated with someone on Micay's blacklist. As he stated, he wants to make his software in peace. Drama could make him cancel the development of the app, and if that happens we're all losing. Especially Patrick Gold.

Anyway, this will be my last post here. I just hope Patrick Gold will keep working on this great keyboard. Maybe Gold and Micay can sort this out, maybe they can't. 

I've donated for a long time to GrapheneOS and stuff like this makes me reconsider.

**Comment by alexiaoff** [comment]
@Dyras 
The problem is not the developers but the GPL3 license. Absolutely all GrapheneOS users would like to have a keyboard like florisboard by default. 
This is not a problem between the developers of GrapheneOS and florisboard but a problem with individuals who try to scuttle both projects. Because once again, the drama would be that florisboard is not implemented in GrapheneOS or another OS.

**Comment by alexiaoff** [comment]
@Dyras 
And if you want a tip keep donating to GrapheneOS. Because thanks to this project ALL Android users gain in security.

**Comment by Ilithy** [comment]
@alexiaoff the fact that Florisboard is not implemented by default in GrapheneOS is not the end of the world, and does not prevent its use on this OS (or other, Calyx for example), why so much passion?

 from a simple user (written with Florisboard on a Pixel 6 pro under Grapheneos) <3

**Comment by popfizzer** [comment]
Pretty shocked to read that thestinger is acting like this. I hope you'll keep working on the project @patrickgold! Use whatever license you want, avoid the drama queens and their hostility!

**Comment by popfizzer** [comment]
@alexiaoff
Then perhaps it should be of the utmost importance that the main developer of GrapheneOS should stay professional with the developers that he is communicating with? 😉

**Comment by KLaurence** [comment]
There seems to be an explosion of activity for this issue. Was this issue linked somewhere outside of GitHub?

**Comment by patrickgold** [comment]
@alexiaoff Please be respectful to others, we want to create a welcoming atmosphere and not a hostile one, this includes accepting that different people have different vies on certain topics.

Also this discussion has gone really off-topic and I don't want this issue to become the start of another drama, so once again my stance:

It does NOT matter for the average end user if FlorisBoard is default in any OS or not, you can easily install FlorisBoard on any Android 7+ phone right now, this includes GrapheneOS, and nothing will change here.  
Including FlorisBoard as a default mostly means more visibility and thus more potential contributors, I agree here, but if this then ends like Bromite nothing is gained from such collab with GrapheneOS. Because sorting out drama means less time spent on actual development, which affects all FlorisBoard users in some way (slower updates, etc.). Also like I said my community should be welcoming and respectful and I want it to stay this way.

I hope that this stance is understandable for everyone and that we can then round off our off-topic course.

**Comment by patrickgold** [comment]
If you want to partake in meaningful discussion why did you just delete your account again @alexiaoff? Extremely unprofessional... Anyways will still respond because I can't let this comment stay as it is. 

1. I never reviewed GPL 3.0 issues, the license issue concludes around the **L**GPL 3.0 license in one of my dependencies. Before accusing others of spreading lies and misinformation get your facts straight please.
2. How can the Bromite drama be a lie when I have a straight up link to the original issue and the original request?! As for other dramas with one OS vs the other I may not have a clear picture but it's probably best to avoid it IMO or act with caution.
3. My **mood** for the open-source world is to be _open and respectful_ and avoid drama, is this such a bad stance in today's world?

I hope that we can now conclude our off-topic track, else I might have to temporarily lock this conversation if this gets out-of-hand.

**Comment by thestinger** [comment]
@patrickgold It's unfortunate you choose to believe and spread baseless libel about myself and GrapheneOS. You've given a platform to abusive trolls spreading baseless lies. You aren't welcome in the GrapheneOS community and FlorisBoard will not be included in GrapheneOS. It will be removed from all GrapheneOS documentation and marked with a warning on every install. Your behavior here and elsewhere has been archived.

I expect you to remove the code contributed to GrapheneOS based on the false pretenses you presented to us which was incredibly dishonest and violation of an informal agreement.

You say you want to avoid drama but engage in abusive behavior and libel yourself. Expect it to be an issue for you going forward anywhere you participate, and you'll find that you're not welcome in any spaces where GrapheneOS has influence.

**Comment by Dyras** [comment]
@thestinger I like that your go-to tactic is to accuse him of libel, instead of explaining your side of the situation and perhaps try to repair things with him to set a good example for future collaborations. Truly astounding.

**Comment by thestinger** [comment]
@Dyras Anyone permitting an abusive troll like yourself to spread completely baseless lies and engage in vicious bullying is not someone who belongs in the open source world.

**Comment by thestinger** [comment]
@Dyras Anyone siding with an abusive troll like yourself to spread completely baseless lies and engage in vicious bullying is not someone who belongs in the open source world.

**Comment by Dyras** [comment]
@thestinger Have you ever considered posting some actual evidence for any of this bullying? It's hard to take stuff like this seriously when we have no idea what the drama is about. If you actually cleared that up for us, perhaps with some screenshots, I'm sure people would be a lot more forgiving when you cut ties with people over the Calyx nonsense.

But yeah, this is off-topic. Hope you keep working on the keyboard Patrick! Cant' wait to use it!

**Comment by thestinger** [comment]
@Dyras: We've repeatedly provided screenshots and logs showing the highly abusive behavior. You're engaging in blatant fabrications and spreading their inaccurate talking points here, You're engaging in libel, harassment and bullying. You are yourself proving what's happening.

**Comment by Dyras** [comment]
@thestinger By all means show those pictures by actually linking them. You can't expect people to go hunting for this on Google, can you?

I'm not proving anything though? If anything you're proving to be incredibly hostile to literally anyone, damaging any future endeavor you might have with other FOSS projects, when instead you could link to what started said drama with Calyx to try to clear things up.

Aren't we supposed to be fighting the big fight together against big tech?

**Comment by thestinger** [comment]
GrapheneOS is no longer interested in using FlorisBoard or making any further contributions to it. Our work improving it was a mistake and a waste of resources. We're thankful we hadn't started doing more of that. There are other keyboard projects we can promote and contribute to instead while discouraging people from using or supporting one with a hostile developer who supports spreading misinformation about our project and developers.

We'll decide on a new path forward and will make an announcement about that as a replacement for the previously announced plans. I'll be posting an announcement retracting our recommendation to use FlorisBoard tomorrow. Since I've regularly promoted it and talked about us planning to use it, that will involve a Twitter post, @room pings on Matrix and other ways of getting that to as many people as possible who have previously seen it.

I've seen enough to know we can't trust this project. Feel free to work with the highly abusive people involved in a harassment and libel campaign targeting us instead. Some people need to learn the hard way that engaging in spreading fabricated stories and bullying people isn't okay and has consequences.

Here's one of the latest examples of the abusive group you've chosen to be part of, sponsored / encouraged by Calyx:

https://twitter.com/DanielMicay/status/1547286521597894657

We've posted many other threads showing examples of the ongoing abusive.

**Comment by popfizzer** [comment]
Yeah uh, that's... F-Droid not Calyx, mate. Might want to consider posting all this on a website or something. Perhaps with a timeline? Obviously the GrapheneOS v CalyxOS is the cause of a lot of friction in the FOSS community, and it doesn't seem like it's getting any better over time.

Just an idea.

**Comment by thestinger** [comment]
@popfizzer Same group of people, largely employed and funded by them. We've posted massive amounts of proof showing the abusive behavior from these groups. You're welcome to go through all the threads about it beginning with the failed Copperhead takeover attempt which Calyx was involved in. The lead developer of CalyxOS was a Copperhead employee and Calyx as an organization supported their attacks to promote their project despite it not being a hardened OS and actually reducing both privacy and especially security compared to AOSP.

**Comment by thestinger** [comment]
@Dyras You're fighting against mobile privacy and security through spreading fabricated stories about GrapheneOS and inaccurate talking points. Your aim is clearly to harm GrapheneOS. Stop pretending you're being constructive when your actions are clearly malicious and aimed at destroying rather than building anything. You've succeeded at cutting off resources from FlorisBoard that we were going to be dedicating to it, and you've succeeded at discouraging us from doing the opposite of continuing to promote it. Siding with people who are openly involved in abuse which we have proven and will continue to proof is not going to go well for the people who choose it. I have zero problem contacting the sponsors, contributors, conferences, employers, partners, etc. of all the people involved in the ongoing misinformation and harassment campaign targeting GrapheneOS.

**Comment by patrickgold** [comment]
@testinger I have no words. Truly no words. So your approach is basically humiliate everyone in the open-source world until your community is completely isolated and at war with everyone? Why do you have to behave this way?

I really expect some kind of professionalism in our work. And if we don't want to work together we just don't work together. But threatening me that it will be an "issue for you going forward anywhere you participate" and that FlorisBoard "will be removed from all GrapheneOS documentation and marked with a warning on every install." is the killer and really too much. Just too much.

**Comment by thestinger** [comment]
@patrickgold Every discussion, every issue, every post about it on other platforms will be used by these malicious trolls to try to attack GrapheneOS at every opportunity. This is every day for FlorisBoard going forward. You've chosen to engage in spreading and supporting fabricated stories, bullying and harassment. It has everything to do with your support for toxic behavior. You chose the wrong side.

FlorisBoard will be marked as coming from a developer involved in the attacks on GrapheneOS. That's completely true and something our community cares about.

Please don't pretend that my behavior is an issue in a thread where you've openly supported these abusive attacks on GrapheneOS and myself after we contributed to your project and promoted it.

**Comment by unilock** [comment]
> @Testinger I have no words.

(you pinged the wrong person)

**Comment by popfizzer** [comment]
Well if it ain't easily accessible and you have to go through years of Twitter for it, can you really expect people to dig through all that? Why not just make a site for it? Or a PDF, or even a Google Doc? 

I haven't heard or read anything about any of this CalyxOS stuff, and at the moment it's hard not to feel that you're coming off as a bully against Florisboard's lead developer. He's one guy doing the world a favor. Why should we hate him for not reading up on the drama?

Keep up the good work @patrickgold! Been loving this app for months now!

**Comment by thestinger** [comment]
@popfizzer Anyone who looks at the actual context of this thread can see that I'm the one who was being targeted by clearly fabricated stories and bullying by malicious trolls, and that @patrickgold supported it. The whole GrapheneOS community will know about it tomorrow if @patrickgold is going to choose that path.

**Comment by popfizzer** [comment]
I dunno man, to me and presumably to most others it seems like you overreacted with the Bromite situation and instead of apologizing and explaining your case you essentially blacklist Patrick's app for something you have to dedicate an evening to fully understanding as it's all convoluted Twitter stuff.

**Comment by patrickgold** [comment]
@thestinger I DO NOT SUPPORT HARASSMENT AND BULLYING. QUIT YOU BS

I am strongly against any kind of harassment against everyone, and your community is the tagert of one, I agree. However harassing other communities, especially communities like FlorisBoard, who have NOTHING TO DO with your ongoing issues is not okay. If you don't want to include FlorisBoard because I am not sure what's going on ok. But please stop spreading lies about me now.

**Comment by thestinger** [comment]
@patrickgold You're still welcome to resolve this in an amicable way where we can go our separate ways instead of what you've decided to do here. I'm more than willing to talk to you on Matrix instead of in a discussion filled with malicious trolls targeting GrapheneOS across platforms.

**Comment by patrickgold** [comment]
@thestinger I agree lets talk this out before we escalate things. What's your username on Matrix?

**Comment by thestinger** [comment]
@strcat:grapheneos.org
